Added a boolean flag for disabling Entity onUpdate

Da-Technomancer opened the pull request:
The flag is false by default, and if true the Entity will not update. It
is saved and loaded from NBT, and is called updateBlocked.
This would Close #3694 and acts in accordance to LexManos's suggestion in the discussion for that issue.

Commit:
Added a boolean flag for disabling Entity onUpdate

The flag is false by default, and if true the Entity will not update. It
is saved and loaded from NBT, and is called updateBlocked.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3731) <br/>All committers have signed the CLA.

williewillus commented:
put a comment so people at know what it does

williewillus commented:
also isn't this going to fail for any entity that overrides onUpdate?

LexManos commented:
Yes which is why the check should be in the place we have the event now, just less crappy event wise.

Da-Technomancer commented:
I don't see what you mean by fail if it is overriden. Yes, overriding the method could stop the flag from having any effect, but overriding the method also can stop the event from having any effect in EntityLivingBase. Also, where should the comment go exactly? On the conditional in onUpdate or the declaration of updateBlocked?

williewillus commented:
imo the flag checking should happen in world / whatever calls the Entity tick, so that it's always checked for every kind of entity, and you can't accidentally get rid of the check by overriding

Commit:
Moved updateBlocked check callers of onUpdate

Also added a comment to the updateBlocked declaration.

Da-Technomancer commented:
This commit adds a comment to the updateBlocked flag, and moves the updateBlocked check to the places that call onUpdate.

mezz commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -493,7 +501,15 @@
+                 }
+                 else
+                 {
++                	if(!p_72866_1_.updateBlocked)
```
you have a tab here

Da-Technomancer committed:
Replaced a tab with 4 spaces

Da-Technomancer commented:
Tab replaced with 4 spaces.

GotoLink commented:
I think this deserve some test case to show it doesn't break other behaviors of entities like loading/unloading, or collision.

Commit:
Added a test case

There is one command added, "blockEntityUpdate" which takes 1 boolean as
a parameter. When used, it sets updateBlocked to the parameter for all
loaded non-player entities in the dimension on the server side.

Da-Technomancer commented:
I've been using the test case. There don't seem to be any problems with loading/unloading or collision that I can find.

HenryLoenwind commented on patches/minecraft/net/minecraft/entity/Entity.java.patch:
```diff
@@ -9,7 +9,15 @@
+     private boolean field_184237_aG;
+     private double[] field_191505_aI;
+     private long field_191506_aJ;
++    public boolean updateBlocked;//Setting this to true will prevent the world from calling onUpdate for this entity. 
```
can you make this a javadoc comment? Then we'll see it on mouseover everwhere this field is used. Or would this make the patch too big?

mezz replied:
Not sure Henry. 
@LexManos what is the right way to document this new field, can it be a javadoc?

LexManos replied:
javadoc is fine,  especially if it's modder facing. Internal stuff with //FIXME, //TODO, or //REMOVE directed twards myself/cpw/fry are fine.

Commit:
Converted comment to JavaDoc

As per HenryLoenwind's request.

Da-Technomancer commented:
Comment replaced with JavaDoc.

mezz commented:
Looks good. Tested every vanilla entity and it works as intended.
There are some interesting side effect of blocking updates since everything is basically frozen in time, like zombies will not burn in daylight, and burning zombies will not take damage. This PR does what it's supposed to though.
